gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -166,6 +168,71 @@ def __ge__(self, other):
 
     return other <= self
 
+  def to_vcf_line(self):
+    """Converts the variant back to a line that can be written to a .vcf."""
+    info_list = []
+    # Set end in info if this is a non-variant.
+    if (self.start and self.reference_bases and
+        self.start + len(self.reference_bases) < self.end):
+      info_list.append('END=%d' % self.end)
+    # Set all other fields of info.
+    for k, v in self.info.iteritems():
+      if v.field_count == '1':
+        info_list.append('%s=%s' % (k, str(v.data)))
+      else:
+        info_list.append('%s=%s' % (k, ','.join([str(x) for x in v.data])))
+    info_field = ';'.join(info_list)
+
+    format_keys = ['GT']
+    alternate_phaseset = False
+    if self.calls:
+      # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+      # add the key to the format field.
+      for call in self.calls:
+        if call.phaseset and call.phaseset != DEFAULT_PHASESET_VALUE:
+          format_keys.append('PS')
+          alternate_phaseset = True
+          break
+      format_keys.extend([k for k in sorted(self.calls[0].info)])
+    format_field = ':'.join(format_keys)
+
+    calls_values = []
+    for call in self.calls:
+      # Set GT field.
+      genotype_field = MISSING_FIELD_VALUE
+      if call.genotype != [MISSING_GENOTYPE_VALUE]:
+        phase_char = '/' if call.phaseset is None else '|'
+        genotype_field = phase_char.join([str(x) for x in call.genotype])
+      call_info_fields = [genotype_field]
+      # Set PS field if there is a set phaseset that is not
+      # `DEFAULT_PHASET_VALUE`.
+      if alternate_phaseset:
+        if call.phaseset and call.phaseset != DEFAULT_PHASESET_VALUE:
+          call_info_fields.append(call.phaseset)
+        else:
+          call_info_fields.append(MISSING_FIELD_VALUE)
+      # Set all other fields of the call info.
+      for k in sorted(call.info):
+        if call.info[k]:
+          call_info_fields.append(str(call.info[k]))
+        else:
+          call_info_fields.append(MISSING_FIELD_VALUE)
+
+      calls_values.append(':'.join(call_info_fields))
+
+    calls_field = '\t'.join(calls_values)
+
+    return '\t'.join(str(x) if x else MISSING_FIELD_VALUE for x in [
+        self.reference_name,
+        self.start + 1 if self.start else None,
+        ';'.join(self.names),
+        self.reference_bases,
+        ','.join(self.alternate_bases),
+        self.quality,
+        ';'.join(self.filters),
+        info_field,
+        format_field]) + ('\t' + calls_field if calls_field else '') + '\n'
```

Comment:
It's cleaner if you have a list of 'columns' where you just return '\t'.join(columns). You can conditionally append things to the list.

Reply:
Done.